File: app/src/main/java/com/miaadrajabi/menuxmlparser/java/data/source/LocalXmlDataSource.java
package com.miaadrajabi.menuxmlparser.java.data.source;

import android.content.Context;
import com.miaadrajabi.menuxmlparser.java.data.xml.XmlModels;
import com.miaadrajabi.menuxmlparser.java.data.xml.XmlConfig;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;

public class LocalXmlDataSource {
    private final Context context;

    public LocalXmlDataSource(Context context) {
        this.context = context;
    }

    private File file(String path) {
        if (path == null) path = DEFAULT_PATH;
        File f = new File(context.getFilesDir(), path);
        File parent = f.getParentFile();
        if (parent != null) {
            parent.mkdirs();
        }
        return f;
    }

    public XmlModels read(String path) throws IOException {
        File f = file(path);
        if (!f.exists()) return null;
        try (FileInputStream input = new FileInputStream(f)) {
            return XmlConfig.getXmlMapper().readValue(input, XmlModels.class);
        }
    }

    public XmlModels read() throws IOException {
        return read(DEFAULT_PATH);
    }

    public void write(XmlModels data, String path) throws IOException {
        File f = file(path);
        try (FileOutputStream output = new FileOutputStream(f)) {
            XmlConfig.getXmlMapper().writeValue(output, data);
        }
    }

    public void write(XmlModels data) throws IOException {
        write(data, DEFAULT_PATH);
    }

    public static final String DEFAULT_PATH = "menus/dynamic_menus.xml";
}

File: app/src/main/java/com/miaadrajabi/menuxmlparser/data/source/LocalXmlDataSource.kt
package com.miaadrajabi.menuxmlparser.data.source

import android.content.Context
import com.miaadrajabi.menuxmlparser.data.xml.DynamicMenusXml
import com.miaadrajabi.menuxmlparser.data.xml.XmlConfig
import java.io.File

class LocalXmlDataSource(private val context: Context) {
    private fun file(path: String = DEFAULT_PATH): File {
        val f = File(context.filesDir, path)
        f.parentFile?.mkdirs()
        return f
    }

    fun read(path: String = DEFAULT_PATH): DynamicMenusXml? {
        val f = file(path)
        if (!f.exists()) return null
        return f.inputStream().use { XmlConfig.xmlMapper.readValue(it, DynamicMenusXml::class.java) }
    }

    fun write(data: DynamicMenusXml, path: String = DEFAULT_PATH) {
        val f = file(path)
        f.outputStream().use { XmlConfig.xmlMapper.writeValue(it, data) }
    }

    companion object {
        const val DEFAULT_PATH = "menus/dynamic_menus.xml"
    }
}


